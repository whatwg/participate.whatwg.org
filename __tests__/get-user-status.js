"use strict";

let mockData = new Map(); // Type -> data file contents as a JS object
let mockOrgMemberships = new Map(); // GitHub user ID -> array of GitHub orgs

jest.mock("../lib/helpers/json-github-database.js", () => {
  return {
    get(type) {
      return {
        sha: "doesn't matter",
        content: mockData.get(type) || []
      };
    }
  };
});

jest.mock("../lib/helpers/github.js", () => {
  return {
    api: {
      orgs: {
        checkPublicMembership({ org, username }) {
          if (mockOrgMemberships.get(username).includes(org)) {
            return Promise.resolve();
          }

          return Promise.reject(new Error("Doesn't matter"));
        }
      }
    }
  };
});


const getUserStatus = require("../lib/get-user-status.js");

function individualData(workstreams, verified, { id = "johndoetw" } = {}) {
  return {
    verified,
    workstreams,
    info: {
      gitHubID: id
    }
  };
}

function entityData(workstreams, verified, { org = "contoso", name = "Contoso Ltd." } = {}) {
  return {
    verified,
    workstreams,
    info: {
      name,
      gitHubOrganization: org
    }
  };
}

beforeEach(() => {
  mockData = new Map();
  mockOrgMemberships = new Map();
});

test("No data", async () => {
  expect(await getUserStatus("johndoetw", "console")).toMatchSnapshot();
});


test("Individual, participating in that particular workstream, verified", async () => {
  mockData.set("individual-public", [individualData(["console"], true)]);
  expect(await getUserStatus("johndoetw", "console")).toMatchSnapshot();
});

test("Individual, participating in that particular workstream, unverified", async () => {
  mockData.set("individual-public", [individualData(["console"], false)]);
  expect(await getUserStatus("johndoetw", "console")).toMatchSnapshot();
});

test("Individual, participating in all workstreams, verified", async () => {
  mockData.set("individual-public", [individualData("all", true)]);
  expect(await getUserStatus("johndoetw", "console")).toMatchSnapshot();
});

test("Individual, participating in all workstreams, unverified", async () => {
  mockData.set("individual-public", [individualData("all", false)]);
  expect(await getUserStatus("johndoetw", "console")).toMatchSnapshot();
});

test("Individual, participating in other workstreams but not that one, verified", async () => {
  mockData.set("individual-public", [individualData(["xhr"], true)]);
  expect(await getUserStatus("johndoetw", "console")).toMatchSnapshot();
});

test("Individual, participating in other workstreams but not that one, unverified", async () => {
  mockData.set("individual-public", [individualData(["xhr"], false)]);
  expect(await getUserStatus("johndoetw", "console")).toMatchSnapshot();
});


test("Via entity, participating in that particular workstream, verified", async () => {
  mockOrgMemberships.set("johndoetw", ["contoso"]);
  mockData.set("entity", [entityData(["console"], true)]);
  expect(await getUserStatus("johndoetw", "console")).toMatchSnapshot();
});

test("Via entity, participating in that particular workstream, unverified", async () => {
  mockOrgMemberships.set("johndoetw", ["contoso"]);
  mockData.set("entity", [entityData(["console"], false)]);
  expect(await getUserStatus("johndoetw", "console")).toMatchSnapshot();
});

test("Via entity, participating in all workstreams, verified", async () => {
  mockOrgMemberships.set("johndoetw", ["contoso"]);
  mockData.set("entity", [entityData("all", true)]);
  expect(await getUserStatus("johndoetw", "console")).toMatchSnapshot();
});

test("Via entity, participating in all workstreams, unverified", async () => {
  mockOrgMemberships.set("johndoetw", ["contoso"]);
  mockData.set("entity", [entityData("all", false)]);
  expect(await getUserStatus("johndoetw", "console")).toMatchSnapshot();
});

test("Via entity, participating in other workstreams but not that one, verified", async () => {
  mockOrgMemberships.set("johndoetw", ["contoso"]);
  mockData.set("entity", [entityData(["xhr"], true)]);
  expect(await getUserStatus("johndoetw", "console")).toMatchSnapshot();
});

test("Via entity, participating in other workstreams but not that one, unverified", async () => {
  mockOrgMemberships.set("johndoetw", ["contoso"]);
  mockData.set("entity", [entityData(["xhr"], false)]);
  expect(await getUserStatus("johndoetw", "console")).toMatchSnapshot();
});


test("Multiple entities, one participating and one not", async () => {
  mockOrgMemberships.set("johndoetw", ["contoso1", "contoso2"]);
  mockData.set("entity", [
    entityData(["xhr"], true, { name: "Contoso 1", org: "contoso1" }),
    entityData(["console"], true, { name: "Contoso 2", org: "contoso2" })
  ]);
  expect(await getUserStatus("johndoetw", "console")).toMatchSnapshot();
});

test("Multiple entities, one verified and one not", async () => {
  mockOrgMemberships.set("johndoetw", ["contoso1", "contoso2"]);
  mockData.set("entity", [
    entityData(["console"], false, { name: "Contoso 1", org: "contoso1" }),
    entityData(["console"], true, { name: "Contoso 2", org: "contoso2" })
  ]);
  expect(await getUserStatus("johndoetw", "console")).toMatchSnapshot();
});

test("Multiple entities, participating one unverified", async () => {
  mockOrgMemberships.set("johndoetw", ["contoso1", "contoso2"]);
  mockData.set("entity", [
    entityData(["console"], false, { name: "Contoso 1", org: "contoso1" }),
    entityData(["xhr"], true, { name: "Contoso 2", org: "contoso2" })
  ]);
  expect(await getUserStatus("johndoetw", "console")).toMatchSnapshot();
});

test("Multiple entities, one good option", async () => {
  mockOrgMemberships.set("johndoetw", ["contoso1", "contoso2"]);
  mockData.set("entity", [
    entityData(["xhr"], false, { name: "Contoso 1", org: "contoso1" }),
    entityData(["console"], true, { name: "Contoso 2", org: "contoso2" })
  ]);
  expect(await getUserStatus("johndoetw", "console")).toMatchSnapshot();
});

test("Multiple entities, all unverified", async () => {
  mockOrgMemberships.set("johndoetw", ["contoso1", "contoso2"]);
  mockData.set("entity", [
    entityData(["console"], false, { name: "Contoso 1", org: "contoso1" }),
    entityData(["console"], false, { name: "Contoso 2", org: "contoso2" })
  ]);
  expect(await getUserStatus("johndoetw", "console")).toMatchSnapshot();
});

test("Multiple entities, all not in this workstream", async () => {
  mockOrgMemberships.set("johndoetw", ["contoso1", "contoso2"]);
  mockData.set("entity", [
    entityData(["xhr"], true, { name: "Contoso 1", org: "contoso1" }),
    entityData(["xhr"], true, { name: "Contoso 2", org: "contoso2" })
  ]);
  expect(await getUserStatus("johndoetw", "console")).toMatchSnapshot();
});


test("Entities exist, but the user is not in them", async () => {
  mockOrgMemberships.set("johndoetw", []);
  mockData.set("entity", [
    entityData(["console"], true, { name: "Contoso 1", org: "contoso1" }),
    entityData(["console"], true, { name: "Contoso 2", org: "contoso2" })
  ]);
  expect(await getUserStatus("johndoetw", "console")).toMatchSnapshot();
});

test("Individuals exist, but the user is not one of them", async () => {
  mockData.set("individual-public", [
    individualData(["console"], true, { id: "janedoetw" }),
    individualData(["console"], true, { id: "bobbosstw" })
  ]);
  expect(await getUserStatus("johndoetw", "console")).toMatchSnapshot();
});

test("Individuals exist, but the user is not one of them; it is an XSS attempt", async () => {
  mockData.set("individual-public", [
    individualData(["console"], true, { id: "janedoetw" }),
    individualData(["console"], true, { id: "bobbosstw" })
  ]);
  expect(await getUserStatus("<script>alert(1);</script>", "console")).toMatchSnapshot();
});
