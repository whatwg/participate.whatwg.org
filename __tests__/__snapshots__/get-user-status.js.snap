// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entities exist, but the user is not in them 1`] = `
Object {
  "context": "Participation",
  "description": "@johndoetw has not signed up to participate",
  "isNothing": true,
  "longDescription": "@johndoetw does not appear to have signed the agreement, or be associated with an entity that has done so. Please sign it, or forward it to an authorized representative!",
  "state": "pending",
  "target_url": "https://participate.whatwg.org/agreement-status?user=johndoetw&repo=console",
}
`;

exports[`Individual, participating in all workstreams, unverified 1`] = `
Object {
  "context": "Participation",
  "description": "@johndoetw is not yet verified",
  "isNothing": false,
  "longDescription": "@johndoetw has signed up to participate as an individual, but has not yet been verified. Hold tight while we sort this out on our end!",
  "state": "pending",
  "target_url": "https://participate.whatwg.org/agreement-status?user=johndoetw&repo=console",
}
`;

exports[`Individual, participating in all workstreams, verified 1`] = `
Object {
  "context": "Participation",
  "description": "@johndoetw has signed up to participate as an individual",
  "isNothing": false,
  "longDescription": "@johndoetw has signed up to participate as an individual. All is well; contribute at will!",
  "state": "success",
  "target_url": "https://participate.whatwg.org/agreement-status?user=johndoetw&repo=console",
}
`;

exports[`Individual, participating in other workstreams but not that one, unverified 1`] = `
Object {
  "context": "Participation",
  "description": "@johndoetw is not yet verified",
  "isNothing": false,
  "longDescription": "@johndoetw has signed up to participate as an individual, but has not yet been verified. Hold tight while we sort this out on our end!",
  "state": "pending",
  "target_url": "https://participate.whatwg.org/agreement-status?user=johndoetw&repo=console",
}
`;

exports[`Individual, participating in other workstreams but not that one, verified 1`] = `
Object {
  "context": "Participation",
  "description": "@johndoetw participates as an individual, but not in this workstream",
  "isNothing": false,
  "longDescription": "@johndoetw has signed up to participate as an individual, but has not chosen to participate in the Console APIs workstream. To amend your agreement, submit a pull request to <a href=\\"https://github.com/whatwg/participant-data\\">whatwg/participant-data</a>.",
  "state": "pending",
  "target_url": "https://participate.whatwg.org/agreement-status?user=johndoetw&repo=console",
}
`;

exports[`Individual, participating in that particular workstream, unverified 1`] = `
Object {
  "context": "Participation",
  "description": "@johndoetw is not yet verified",
  "isNothing": false,
  "longDescription": "@johndoetw has signed up to participate as an individual, but has not yet been verified. Hold tight while we sort this out on our end!",
  "state": "pending",
  "target_url": "https://participate.whatwg.org/agreement-status?user=johndoetw&repo=console",
}
`;

exports[`Individual, participating in that particular workstream, verified 1`] = `
Object {
  "context": "Participation",
  "description": "@johndoetw has signed up to participate as an individual",
  "isNothing": false,
  "longDescription": "@johndoetw has signed up to participate as an individual. All is well; contribute at will!",
  "state": "success",
  "target_url": "https://participate.whatwg.org/agreement-status?user=johndoetw&repo=console",
}
`;

exports[`Individuals exist, but the user is not one of them 1`] = `
Object {
  "context": "Participation",
  "description": "@johndoetw has not signed up to participate",
  "isNothing": true,
  "longDescription": "@johndoetw does not appear to have signed the agreement, or be associated with an entity that has done so. Please sign it, or forward it to an authorized representative!",
  "state": "pending",
  "target_url": "https://participate.whatwg.org/agreement-status?user=johndoetw&repo=console",
}
`;

exports[`Individuals exist, but the user is not one of them; it is an XSS attempt 1`] = `
Object {
  "context": "Participation",
  "description": "@<script>alert(1);</script> has not signed up to participate",
  "isNothing": true,
  "longDescription": "@&lt;script&gt;alert(1);&lt;/script&gt; does not appear to have signed the agreement, or be associated with an entity that has done so. Please sign it, or forward it to an authorized representative!",
  "state": "pending",
  "target_url": "https://participate.whatwg.org/agreement-status?user=<script>alert(1);</script>&repo=console",
}
`;

exports[`Multiple entities, all not in this workstream 1`] = `
Object {
  "context": "Participation",
  "description": "@johndoetw participates on behalf of a non-workstream entity",
  "isNothing": false,
  "longDescription": "@johndoetw is part of the contoso1 and contoso2 GitHub organizations, but they all do not participate in the Console APIs workstream. Please have the contact for those organizations update their participation agreement.",
  "state": "pending",
  "target_url": "https://participate.whatwg.org/agreement-status?user=johndoetw&repo=console",
}
`;

exports[`Multiple entities, all unverified 1`] = `
Object {
  "context": "Participation",
  "description": "@johndoetw participates on behalf of an unverified entity",
  "isNothing": false,
  "longDescription": "@johndoetw is part of the contoso1 and contoso2 GitHub organizations, but they all have not yet been verified. Hold tight while we sort this out on our end!",
  "state": "pending",
  "target_url": "https://participate.whatwg.org/agreement-status?user=johndoetw&repo=console",
}
`;

exports[`Multiple entities, one good option 1`] = `
Object {
  "context": "Participation",
  "description": "johndoetw participates on behalf of Contoso 2",
  "isNothing": false,
  "longDescription": "@johndoetw is part of the contoso2 GitHub organization, associated with Contoso 2, which has signed the agreement and participates in the Console APIs workstream. All is well; contribute at will!",
  "state": "success",
  "target_url": "https://participate.whatwg.org/agreement-status?user=johndoetw&repo=console",
}
`;

exports[`Multiple entities, one participating and one not 1`] = `
Object {
  "context": "Participation",
  "description": "johndoetw participates on behalf of Contoso 2",
  "isNothing": false,
  "longDescription": "@johndoetw is part of the contoso2 GitHub organization, associated with Contoso 2, which has signed the agreement and participates in the Console APIs workstream. All is well; contribute at will!",
  "state": "success",
  "target_url": "https://participate.whatwg.org/agreement-status?user=johndoetw&repo=console",
}
`;

exports[`Multiple entities, one verified and one not 1`] = `
Object {
  "context": "Participation",
  "description": "johndoetw participates on behalf of Contoso 2",
  "isNothing": false,
  "longDescription": "@johndoetw is part of the contoso2 GitHub organization, associated with Contoso 2, which has signed the agreement and participates in the Console APIs workstream. All is well; contribute at will!",
  "state": "success",
  "target_url": "https://participate.whatwg.org/agreement-status?user=johndoetw&repo=console",
}
`;

exports[`Multiple entities, participating one unverified 1`] = `
Object {
  "context": "Participation",
  "description": "@johndoetw participates on behalf of a non-workstream entity",
  "isNothing": false,
  "longDescription": "@johndoetw is part of the contoso2 GitHub organization, but it does not participate in the Console APIs workstream. Please have the contact for this organization update their participation agreement.",
  "state": "pending",
  "target_url": "https://participate.whatwg.org/agreement-status?user=johndoetw&repo=console",
}
`;

exports[`No data 1`] = `
Object {
  "context": "Participation",
  "description": "@johndoetw has not signed up to participate",
  "isNothing": true,
  "longDescription": "@johndoetw does not appear to have signed the agreement, or be associated with an entity that has done so. Please sign it, or forward it to an authorized representative!",
  "state": "pending",
  "target_url": "https://participate.whatwg.org/agreement-status?user=johndoetw&repo=console",
}
`;

exports[`Via entity, participating in all workstreams, unverified 1`] = `
Object {
  "context": "Participation",
  "description": "@johndoetw participates on behalf of an unverified entity",
  "isNothing": false,
  "longDescription": "@johndoetw is part of the contoso GitHub organization, but it has not yet been verified. Hold tight while we sort this out on our end!",
  "state": "pending",
  "target_url": "https://participate.whatwg.org/agreement-status?user=johndoetw&repo=console",
}
`;

exports[`Via entity, participating in all workstreams, verified 1`] = `
Object {
  "context": "Participation",
  "description": "johndoetw participates on behalf of Contoso Ltd.",
  "isNothing": false,
  "longDescription": "@johndoetw is part of the contoso GitHub organization, associated with Contoso Ltd., which has signed the agreement and participates in the Console APIs workstream. All is well; contribute at will!",
  "state": "success",
  "target_url": "https://participate.whatwg.org/agreement-status?user=johndoetw&repo=console",
}
`;

exports[`Via entity, participating in other workstreams but not that one, unverified 1`] = `
Object {
  "context": "Participation",
  "description": "@johndoetw participates on behalf of an unverified entity",
  "isNothing": false,
  "longDescription": "@johndoetw is part of the contoso GitHub organization, but it has not yet been verified. Hold tight while we sort this out on our end!",
  "state": "pending",
  "target_url": "https://participate.whatwg.org/agreement-status?user=johndoetw&repo=console",
}
`;

exports[`Via entity, participating in other workstreams but not that one, verified 1`] = `
Object {
  "context": "Participation",
  "description": "@johndoetw participates on behalf of a non-workstream entity",
  "isNothing": false,
  "longDescription": "@johndoetw is part of the contoso GitHub organization, but it does not participate in the Console APIs workstream. Please have the contact for this organization update their participation agreement.",
  "state": "pending",
  "target_url": "https://participate.whatwg.org/agreement-status?user=johndoetw&repo=console",
}
`;

exports[`Via entity, participating in that particular workstream, unverified 1`] = `
Object {
  "context": "Participation",
  "description": "@johndoetw participates on behalf of an unverified entity",
  "isNothing": false,
  "longDescription": "@johndoetw is part of the contoso GitHub organization, but it has not yet been verified. Hold tight while we sort this out on our end!",
  "state": "pending",
  "target_url": "https://participate.whatwg.org/agreement-status?user=johndoetw&repo=console",
}
`;

exports[`Via entity, participating in that particular workstream, verified 1`] = `
Object {
  "context": "Participation",
  "description": "johndoetw participates on behalf of Contoso Ltd.",
  "isNothing": false,
  "longDescription": "@johndoetw is part of the contoso GitHub organization, associated with Contoso Ltd., which has signed the agreement and participates in the Console APIs workstream. All is well; contribute at will!",
  "state": "success",
  "target_url": "https://participate.whatwg.org/agreement-status?user=johndoetw&repo=console",
}
`;
